File: tokenizer.py
import re
from collections import Counter

class AllPurposeTokenizer:
    def __init__(self):
        # GPT-4 inspired regex pattern to handle whitespace, letters, numbers, and punctuation
        self.pat = re.compile(r"""'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+""")
        self.merges = {}  # Stores pair:new_token mappings
        self.vocab = {}   # Stores token_id:byte_value mappings

    def get_stats(self, ids):
        """Counts the frequency of adjacent pairs in a list of IDs."""
        counts = Counter()
        for pair in zip(ids, ids[1:]):
            counts[pair] += 1
        return counts

    def merge(self, ids, pair, idx):
        """Replaces all occurrences of a pair with a new merged token ID."""
        new_ids = []
        i = 0
        while i < len(ids):
            if i < len(ids) - 1 and ids[i] == pair[0] and ids[i+1] == pair[1]:
                new_ids.append(idx)
                i += 2
            else:
                new_ids.append(ids[i])
                i += 1
        return new_ids

    def train(self, text, vocab_size, verbose=False):
        """Trains the BPE tokenizer on a given corpus."""
        num_merges = vocab_size - 256
        # Initial tokenization: convert text to UTF-8 bytes
        tokens = list(text.encode("utf-8"))
        ids = list(tokens)
        
        for i in range(num_merges):
            stats = self.get_stats(ids)
            if not stats:
                break
            # Find the most frequent pair
            top_pair = max(stats, key=stats.get)
            idx = 256 + i
            if verbose:
                print(f"Merging {top_pair} into new token {idx}")
            ids = self.merge(ids, top_pair, idx)
            self.merges[top_pair] = idx
            
    def encode(self, text):
        """Encodes raw text into token IDs."""
        # First, split the text using the regex pattern
        text_chunks = re.findall(self.pat, text)
        all_ids = []
        
        for chunk in text_chunks:
            chunk_ids = list(chunk.encode("utf-8"))
            # Iteratively apply learned merges
            while len(chunk_ids) >= 2:
                stats = self.get_stats(chunk_ids)
                # Find the merge that occurred earliest in training
                pair = min(stats, key=lambda p: self.merges.get(p, float("inf")))
                if pair not in self.merges:
                    break
                idx = self.merges[pair]
                chunk_ids = self.merge(chunk_ids, pair, idx)
            all_ids.extend(chunk_ids)
        return all_ids

    def decode(self, ids):
        """Decodes token IDs back into a string."""
        # Convert IDs back to bytes based on initial 0-255 map + merges
        # Note: A full implementation would build a reverse vocab map during training
        tokens = b"".join([bytes([idx]) if idx < 256 else b"" for idx in ids])
        return tokens.decode("utf-8", errors="replace")

File: src/core/tokenizer.ts
// Ported from the Lab 01 tokenizer concepts (BPE + regex splitting).

export type MergePair = `${number},${number}`;

function pairKey(a: number, b: number): MergePair {
  return `${a},${b}`;
}

function bytesToString(bytes: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(bytes);
}

function stringToBytes(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

export class BPETokenizer {
  // Simple BPE on bytes. Includes special tokens similar to the lab.
  merges: Map<MergePair, number>;
  vocab: Map<number, Uint8Array>;
  specialTokens: Map<string, number>;
  private inverseSpecialTokens: Map<number, string>;

  constructor() {
    this.merges = new Map();
    this.vocab = new Map();

    for (let i = 0; i < 256; i++) this.vocab.set(i, new Uint8Array([i]));

    // Match the Lab 01 convention: <|sos|> at 256, <|eos|> at 257
    this.specialTokens = new Map([
      ["<|sos|>", 256],
      ["<|eos|>", 257],
      ["<|unk|>", 258],
    ]);
    this.inverseSpecialTokens = new Map();
    for (const [k, v] of this.specialTokens.entries()) {
      this.vocab.set(v, stringToBytes(k));
      this.inverseSpecialTokens.set(v, k);
    }
  }

  private getStats(ids: number[]): Map<MergePair, number> {
    const counts = new Map<MergePair, number>();
    for (let i = 0; i < ids.length - 1; i++) {
      const k = pairKey(ids[i], ids[i + 1]);
      counts.set(k, (counts.get(k) ?? 0) + 1);
    }
    return counts;
  }

  private mergePairs(ids: number[], pair: MergePair, newId: number): number[] {
    const [aStr, bStr] = pair.split(",");
    const a = Number(aStr);
    const b = Number(bStr);
    const out: number[] = [];
    let i = 0;
    while (i < ids.length) {
      if (i < ids.length - 1 && ids[i] === a && ids[i + 1] === b) {
        out.push(newId);
        i += 2;
      } else {
        out.push(ids[i]);
        i += 1;
      }
    }
    return out;
  }

  train(text: string, maxVocabSize: number, verbose = false): void {
    const pretrainSize = this.vocab.size;
    const numMerges = Math.max(0, maxVocabSize - pretrainSize);
    let ids = Array.from(stringToBytes(text), (b) => b);

    for (let i = 0; i < numMerges; i++) {
      const stats = this.getStats(ids);
      if (stats.size === 0) break;

      // Find most frequent pair (Lab 01 simple tokenizer behavior)
      let bestPair: MergePair | null = null;
      let bestCount = -1;
      for (const [p, c] of stats.entries()) {
        if (c > bestCount) {
          bestPair = p;
          bestCount = c;
        }
      }
      if (!bestPair) break;

      const newId = this.vocab.size;
      this.merges.set(bestPair, newId);

      const [aStr, bStr] = bestPair.split(",");
      const a = Number(aStr);
      const b = Number(bStr);
      const aBytes = this.vocab.get(a);
      const bBytes = this.vocab.get(b);
      if (!aBytes || !bBytes) break;
      const merged = new Uint8Array(aBytes.length + bBytes.length);
      merged.set(aBytes, 0);
      merged.set(bBytes, aBytes.length);
      this.vocab.set(newId, merged);

      ids = this.mergePairs(ids, bestPair, newId);
      if (verbose && i % 250 === 0) {
        console.log(`merge ${i + 1}/${numMerges}: ${bestPair} -> ${newId} (count=${bestCount})`);
      }
    }
  }

  encode(text: string): number[] {
    // Supports special tokens literally present in the text.
    // Strategy: split on special tokens, encode others as bytes + BPE merges.
    if (text.length === 0) return [];

    const specials = Array.from(this.specialTokens.keys());
    const specialRegex = specials.length
      ? new RegExp(`(${specials.map((s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|")})`, "g")
      : null;

    const parts = specialRegex ? text.split(specialRegex).filter((p) => p.length > 0) : [text];
    const out: number[] = [];

    for (const part of parts) {
      const sid = this.specialTokens.get(part);
      if (sid !== undefined) {
        out.push(sid);
        continue;
      }

      // Start from raw bytes
      let ids = Array.from(stringToBytes(part), (b) => b);

      // Iteratively apply the earliest merge available (Lab 01 behavior for encode)
      while (ids.length >= 2) {
        const stats = this.getStats(ids);
        let bestPair: MergePair | null = null;
        let bestMergeId = Infinity;

        for (const p of stats.keys()) {
          const mid = this.merges.get(p);
          if (mid !== undefined && mid < bestMergeId) {
            bestMergeId = mid;
            bestPair = p;
          }
        }

        if (!bestPair || bestMergeId === Infinity) break;
        ids = this.mergePairs(ids, bestPair, bestMergeId);
      }

      out.push(...ids);
    }

    return out;
  }

  decode(ids: number[]): string {
    // Reconstruct bytes from vocab. Unknown ids become <|unk|>.
    const parts: Uint8Array[] = [];
    for (const id of ids) {
      const b = this.vocab.get(id);
      if (b) {
        parts.push(b);
      } else {
        parts.push(stringToBytes(this.inverseSpecialTokens.get(this.specialTokens.get("<|unk|>")!) ?? "<|unk|>"));
      }
    }
    const total = parts.reduce((s, b) => s + b.length, 0);
    const joined = new Uint8Array(total);
    let off = 0;
    for (const b of parts) {
      joined.set(b, off);
      off += b.length;
    }
    return bytesToString(joined);
  }
}

export class RegexBPETokenizer {
  // A more “lab accurate” variant: split with a regex first, then BPE within each chunk.
  private bpe: BPETokenizer;
  private pattern: RegExp;

  constructor() {
    this.bpe = new BPETokenizer();
    // A pragmatic browser-safe regex that splits into (optional leading space + word/number) or punctuation.
    // (The notebook mentions a more GPT‑4-ish regex; JS doesn't support \p{L} on all environments without /u.)
    this.pattern = /'s|'t|'re|'ve|'m|'ll|'d| ?[A-Za-z]+| ?[0-9]+| ?[^ \t\r\nA-Za-z0-9]+|\s+(?!\S)|\s+/g;
  }

  get merges(): Map<MergePair, number> {
    return this.bpe.merges;
  }
  get vocab(): Map<number, Uint8Array> {
    return this.bpe.vocab;
  }
  get specialTokens(): Map<string, number> {
    return this.bpe.specialTokens;
  }

  train(text: string, maxVocabSize: number): void {
    // Train on raw bytes for simplicity (still consistent with Lab 01 idea).
    this.bpe.train(text, maxVocabSize, false);
  }

  encode(text: string): number[] {
    const chunks = text.match(this.pattern) ?? [];
    const out: number[] = [];
    for (const ch of chunks) out.push(...this.bpe.encode(ch));
    return out;
  }

  decode(ids: number[]): string {
    return this.bpe.decode(ids);
  }
}

